### ISSUE TITLE
Race condition in RDIP::Connection with multiple server responses.

### ISSUE THREAD

**Issue by seraku24** [SketchUp/sketchup-ruby-debugger#10]
When `RDIP::Connection` handles multiple commands that require passing control back to the server thread via `server_response_`, subsequent commands have an opportunity to overwrite the previous ones.

I noticed this behavior while debugging `vscode-ruby` to help investigate rubyide/vscode-ruby#95.  When using PuTTY to connect to the debugger and issue commands directly, I never experience a problem, most likely due to the fact that manually typing commands takes a significantly longer time compared to an automated process.

When `vscode-ruby` processes a breakpoint or step, it submits four requests: `thread list`, `where`, `var local`, and `var global`.  Of note, the two variable requests are sent immediately one after the other, without waiting for a response between.  It is these two requests that `RDIP::Connection` needs to forward to the server thread.  And it would seem that, on my computer at least, it is quite often the case that the `var local` command gets overwritten before the server thread can handle it.  This results in the debugger only sending back the global variables.

For reference, if I modify `vscode-ruby` to wait for the local variables before requesting the globals, I do not experience the issue.

The best solution likely is to use a queue of some form.  That said, something as simple as the following probably could work in the various locations the pattern appears:

```cpp
while (server_response_) { /* Do nothing */ }
server_response_ = std::bind( ... );
process_server_response_ = std::bind( ... );
server_wait_cond_.notify_all();
```

**Comment by thomthom** [comment]
@BugraB, @birsoyo any thoughts on this one?

**Comment by BugraB** [comment]
Hi there. I will take a look at this as soon as I get a chance. I haven't been in this code for a while so I need to get reacclimated :) Also, feel free to submit a pull request.
Thanks!

**Comment by seraku24** [comment]
Unfortunately, I do not have a C++ build environment on my system at the moment.  Without being able to build and test any proposed changes, I would not feel comfortable submitting a pull request.

But seeing as I have been meaning to get back into C++ for a number of years now to catch up on all of the features C++11 added, I will try to find some time today to get my Visual Studio updated so I can help here.

**Comment by BugraB** [comment]
Sounds good and thank you!
I will take a look at this but can't promise how soon I'll be able to get to it.

**Comment by thomthom** [comment]
I'm in the process of writing up a guide for debugging with RubyMine. But I also wanted to do one for VSCode. You might have saved me a good amount of pondering frustration by flushing out this issue.

**Comment by seraku24** [comment]
On a happy note, I got the project building and running, albeit needing a few modifications to the Visual Studio project files.  For having been away from C++, I am surprised how much is coming back.  I even still have muscle memory for some Visual Studio shortcuts I have not used in years, which was quite unexpected.

But sadly, **the workaround I proposed does not work reliably.**  `):`  It would seem there are some deeper threading issues at play than I first realized.  Now having reviewed more of the code in action, I foresee a fair bit of refactoring in our future to resolve this robustly.

I will definitely keep poking at this over the next few days, and hopefully I can manage to get something working without having to resort to any drastic, major rewrites.  *\*fingers crossed\** `(:`
